### PR TITLE
Fixed #32916 -- Combined request.META["CSRF_COOKIE_USED"] and request.csrf_cookie_needs_reset.

### PR DESCRIPTION
**Pull request by cjerdonek**
https://code.djangoproject.com/ticket/32916

### CODE REVIEW

File: docs/releases/4.0.txt [hunk 1]
```diff
@@ -191,6 +191,11 @@ CSRF
   this, :ref:`some changes <csrf-trusted-origins-changes-4.0>` to the
   :setting:`CSRF_TRUSTED_ORIGINS` setting are required.
 
+* :class:`~django.middleware.csrf.CsrfViewMiddleware` now uses
+  ``request.META["CSRF_COOKIE_NEEDS_RESET"]`` in place of
+  ``request.META["CSRF_COOKIE_USED"]``, ``request.csrf_cookie_needs_reset``, and
+  ``request.csrf_cookie_set``to track whether the CSRF cookie should be sent.
```
Comment (cjerdonek) [inline docs/releases/4.0.txt:197]: Should there be some wording here to signal that these are internal API's?

File: docs/releases/4.0.txt [hunk 2]
```diff
@@ -191,6 +191,11 @@ CSRF
   this, :ref:`some changes <csrf-trusted-origins-changes-4.0>` to the
   :setting:`CSRF_TRUSTED_ORIGINS` setting are required.
 
+* :class:`~django.middleware.csrf.CsrfViewMiddleware` now uses
+  ``request.META["CSRF_COOKIE_NEEDS_RESET"]`` in place of
+  ``request.META["CSRF_COOKIE_USED"]``, ``request.csrf_cookie_needs_reset``, and
+  ``request.csrf_cookie_set`` to track whether the CSRF cookie should be sent.
```
Comment (cjerdonek) [inline docs/releases/4.0.txt:197]: Should there be some wording here to signal that these are internal API's?

**Comment by cjerdonek** [comment]
cc @apollo13 @shaib

**Comment by apollo13** [comment]
In principle I like the change. I am just not sure if we should rename the META flag in the process. I mean there are probably not many users of it (?) but it still "feels"  a bit like an unnecessarily change.

**Comment by cjerdonek** [comment]
Here are a couple reasons in favor of changing it. First, if we keep the existing `"CSRF_COOKIE_USED"` string, then it will no longer be a correct description of what the flag represents. This is because there will now be three cases where the flag is set without the cookie being "used" (in the sense of the token being used in the template, which is what it meant before). That mismatch will be there as long as it's not changed, so it seems better to change it sooner rather than later. Another reason is that if someone is relying on this internal API, they will be less likely to notice the change in behavior / semantics if the name is kept, than if the name is changed. Also, if we wanted to further alert people who may be relying on this internal flag by setting it, we could check for the presence of the old key in `process_response()` and raise an exception if it's found.

**Comment by apollo13** [comment]
Fair enough, but maybe we should call it `_RENEW`. `_RESET` for me seems to imply a more drastical reset like when regenerating a token secret. What do you think?

**Comment by cjerdonek** [comment]
Yeah, I think `RENEW` is probably better. I'm open to brainstorming more in fact if people have other ideas. I mostly just went with the attribute name without thinking too hard.

**Comment by cjerdonek** [comment]
> Given that the flag is used for different cases it might be hard finding a proper matching verb

I think as long as it's viewed as an instruction of _what_ needs to be done rather than why, I think it should be okay. What about any of these? I've limited to names starting with `CSRF_COOKIE` so it will sort next to the cookie value (key `CSRF_COOKIE`) when viewing the `META` dict:

* `CSRF_COOKIE_DO_RENEW`
* `CSRF_COOKIE_DO_SEND`
* `CSRF_COOKIE_DO_SET`
* `CSRF_COOKIE_NEEDS_RENEW`
* `CSRF_COOKIE_NEEDS_SEND`
* `CSRF_COOKIE_NEEDS_SET`
* `CSRF_COOKIE_RENEW`
* `CSRF_COOKIE_SEND`
* `CSRF_COOKIE_SET` (This one would be ambiguous but including for completeness.)

The usage will be here in `process_response()`:

```python
if request.META.get("CSRF_COOKIE_NEEDS_SET"):
    self._set_token(request, response)
    request.META["CSRF_COOKIE_NEEDS_SET"] = False
...
```

**Comment by apollo13** [comment]
What about `CSRF_COOKIE_NEEDS_UPDATE`? It is slightly wrong in the case when we initially set it, but still sounds like a good candidate.

**Comment by cjerdonek** [comment]
> What about CSRF_COOKIE_NEEDS_UPDATE?

I like it! 👍

**Comment by cjerdonek** [comment]
Updated.

**Comment by felixxm** [comment]
@cjerdonek Thanks :+1:

File: docs/releases/4.0.txt (hunk 2 above)
Comment (felixxm) [reply docs/releases/4.0.txt]: I added a note.

**Review by shaib — COMMENTED** [review]
Sorry for getting here only this late; anyway, you seem to have managed fine without me :)

+1

**Pull request merged**